arm_gdb/mpu.py

```diff
@@ -0,0 +1,272 @@
+                ],"Access permissions"),
+            FieldBitfieldEnum("XN", 0, 1, [
+                (0b1, False, "Disallow program execution in this region", None),
+                (0b0, False, "Allow program execution in this region", None),
+                ],"eXecute Never attribute"),
+            ]),
+    ]
+
+def get_rlar_reg():
+    return [
+        RegisterDef("MPU_RLAR", "MPU Region Base Limit Register", 0xE000EDA0, 4, [
+            FieldBitfield("LIMIT", 5, 27, "Ending address (upper inclusive limit) of MPU region address"),
+            FieldBitfieldEnum("PXN", 4, 1, [
+                (0b1, False, "Execution from a privileged mode is not permitted", None),
+                (0b0, False, "Execution only permitted if read permitted", None),
+                ],"Privileged execute-never"),
+            FieldBitfield("AttrIndx", 1, 3, "Attribute Index. Select memory attributes from attribute sets in MPU_MAIR0 and MPU_MAIR1"),
+            FieldBitfieldEnum("EN", 0, 1, [
+                (0b0, False, "disable", None),
+                (0b1, False, "Region enable", None),
+            ],"Region enable"),
+        ]),
+    ]
+
+def get_mpu_region_start():
+    REG_RBAR = 0xE000ED9C
+    value_str = gdb.execute(f"x/1x {REG_RBAR}", to_string=True)
+    value_hex = value_str.split(':')[1].strip()
+    value = int(value_hex, 16)
+    start = value & 0xFFFFFFE0;
+    return start
+
+def get_mpu_region_limit():
+    REG_RLAR = 0xE000EDA0
+    value_str = gdb.execute(f"x/1x {REG_RLAR}", to_string=True)
+    value_hex = value_str.split(':')[1].strip()
+    value = int(value_hex, 16)
+    limit = value & 0xFFFFFFE0;
+    return limit
+
+def mpu_region_enable():
+    REG_RLAR = 0xE000EDA0
+    value_str = gdb.execute(f"x/1x {REG_RLAR}", to_string=True)
+    value_hex = value_str.split(':')[1].strip()
+    value = int(value_hex, 16)
+    enabled = bool(value & 0x01)
+    return enabled
+
+def get_mpu_regions():
+    set_rnr_reg(1)
+    reg = get_rbar_reg()
+    reg.dump(inf, args['descr'], base=base, all=args['all'])
+
+    reg = get_rlar_reg()
+    reg.dump(inf, args['descr'], base=base, all=args['all'])
+
+class ArmToolsMPU (ArgCommand):
+    """Dump of ARM Cortex-M MPU - SCB registers for the FP extension
+
+Usage: arm mpu [/hab]
+
+Modifier /h provides descriptions of names where available
+Modifier /a Print all fields, including default values
+Modifier /b prints bitmasks in binary instead of hex
+"""
+
+    def __init__(self):
+        super().__init__('arm mpu', gdb.COMMAND_DATA)
+        self.add_mod('h', 'descr')
+        self.add_mod('a', 'all')
+        self.add_mod('b', 'binary')
+
+    def invoke(self, argument, from_tty):
+        args = self.process_args(argument)
+        if args is None:
+            self.print_help()
+            return
+
+        base = 1 if args['binary'] else 4
+
+        inf = gdb.selected_inferior()
+
+        print("SCB MPU registers:")
+
+        try:
```

Comment:
For what I understand, MPU is not available on all architectures, and some have them implementation dependent.

This implementation is, for what I can tell, for the ARMv8-M architecture.

I think there should be a check somewhere here that verifies the architecture, and shows a message that it's not supported for the current architecture, if it is something else.

Reply:
Just before I want to use CPUID to difference the arch, but there are some confused infomation about CPUID,
for armv7m, the Architecture bits(16:19) is 0xf, but for armv8-m with Main Extension, it is also 0xf.

So dont have a good idea how to difference the arch.